### PR TITLE
[ML] Time series segmentation improvements

### PR DESCRIPTION
**Pull request by tveasey**
This follows on from #1530 and prepares the ground for using the code to segment time windows of values into piecewise linear trend and piecewise constant scaling for change detection.

Aside from some small improvements to methodology, the main changes were to support scaling an arbitrary seasonal function, which will be the current time series model when using for change point detection, and being able to limit the number of changes we fit. To do this it was useful to rewrite segmentation to use common functionality in `CSignal`.

I'll make the changes to actually use this in a follow up pull request. As before, this will not documented separately so marking as a non-issue.

### CODE REVIEW

**Review by valeriy42 — COMMENTED** [review]
LGTM. Just a few minor comments.

File: lib/maths/CTimeSeriesTestForSeasonality.cc [hunk 1]
```diff
@@ -499,29 +501,18 @@ CSeasonalDecomposition CTimeSeriesTestForSeasonality::select(TModelVec& decompos
...
-
-            // It is possible that the null hypothesis uses a piecewise linear fit of
-            // seasonal components in the data. In this case we accept the alternative
-            // if it's autocorrelation is high and number of segments in the trend
-            // is large enough.
-            if (pValueVsH0 > pValueToAccept &&
-                (fuzzyGreaterThan(logPValue / logAcceptedFalsePositiveRate, 1.0, 1.0) &&
-                 fuzzyGreaterThan(autocorrelation / m_HighAutocorrelation, 1.0, 0.1) &&
-                 fuzzyLessThan(8.0 / decompositions[H0].numberParameters(), 1.0, 0.2))
-                        .boolean() == false) {
```
[the remark this answers is not in the document]
Comment (tveasey) [reply lib/maths/CTimeSeriesTestForSeasonality.cc:524]: Yes. This was introduced to deal with the fact that we could fit seasonality with piecewise linear trend if we used enough segments. Now we explicitly limit the number of trend segments to something small and choose the most significant w.r.t. that constraint this is no longer a problem.

File: lib/maths/CTimeSeriesTestForSeasonality.cc [hunk 2]
```diff
@@ -1285,18 +1298,17 @@ void CTimeSeriesTestForSeasonality::removeDiscontinuities(const TSizeVec& modelT
     }
 }
 
-bool CTimeSeriesTestForSeasonality::meanScale(const SHypothesisStats& hypothesis,
-                                              const TIndexWeight& weight,
+bool CTimeSeriesTestForSeasonality::meanScale(const TSeasonalComponentVec& periods,
+                                              const TSizeVec& scaleSegments,
```
[the remark this answers is not in the document]
Comment (tveasey) [reply lib/maths/CTimeSeriesTestForSeasonality.cc:1298]: Unfortunately, `weight` is a defaulted so I have to keep it at the end.

File: lib/maths/CTimeSeriesTestForSeasonality.cc [hunk 3]
```diff
@@ -565,20 +556,24 @@ CSeasonalDecomposition CTimeSeriesTestForSeasonality::select(TModelVec& decompos
...
                 static_cast<double>(decompositions[H1].s_NumberTrendParameters)};
-            double pValueVsSelected{selected < decompositions.size()
-                                        ? decompositions[H1].pValue(decompositions[selected])
-                                        : 0.0};
+            double pValueH1VsSelected{selected < decompositions.size()
+                                          ? decompositions[H1].pValue(decompositions[selected])
+                                          : 0.0};
+            double pValueSelectedVsH1{selected < decompositions.size()
+                                          ? decompositions[selected].pValue(decompositions[H1])
+                                          : 1.0};
```
[the remark this answers is not in the document]
Comment (tveasey) [reply lib/maths/CTimeSeriesTestForSeasonality.cc]: Done in [37c263e](https://github.com/elastic/ml-cpp/pull/1543/commits/37c263eda2f74ff0baac9c270237796512848a23).

File: lib/maths/unittest/CTimeSeriesSegmentationTest.cc [hunk 4]
```diff
@@ -224,82 +223,62 @@ BOOST_AUTO_TEST_CASE(testPiecewiseLinearScaledSeasonal) {
...
+        LOG_DEBUG(<< "true segmentation = "
+                  << core::CContainerPrinter::print(trueSegmentation));
+        LOG_DEBUG(<< "segmentation      = " << core::CContainerPrinter::print(segmentation));
 
-            // Not biased.
-            BOOST_TEST_REQUIRE(
-                std::fabs(maths::CBasicStatistics::mean(residualMoments) -
-                          maths::CBasicStatistics::mean(noiseMoments)) <
-                3.0 * std::sqrt(maths::CBasicStatistics::variance(noiseMoments) /
-                                maths::CBasicStatistics::count(noiseMoments)));
```
[the remark this answers is not in the document]
Comment (tveasey) [reply lib/maths/unittest/CTimeSeriesSegmentationTest.cc:287]: The reason was that this test was additionally fitting and removing piecewise scaled component. Currently, we weren't using that functionality outside of the tests. I do plan to use it when with the reworking of change detection, but I think it is a better separation of concerns to have a separate test. I'll add that test along with that change.

**Review by valeriy42 — APPROVED** [review]
LGTM

**Pull request merged**